File: src/unicode.pas
{
        Copyright (c) 2020 Kristofer Munsterhjelm

        Permission is hereby granted, free of charge, to any person obtaining a copy
        of this software and associated documentation files (the "Software"), to deal
        in the Software without restriction, including without limitation the rights
        to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
        copies of the Software, and to permit persons to whom the Software is
        furnished to do so, subject to the following conditions:

        The above copyright notice and this permission notice shall be included in all
        copies or substantial portions of the Software.

        THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
        IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
        FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
        AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
        LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
        OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
        SOFTWARE.
}

{ So as not to have to pull in a whole mathematics library just for min and
  max. }

unit Unicode;

interface
    const
        { Credits go to Ben Russell (iamgreaser) for this table and
          the logic behind WriteUnicodeAsUTF8. }
        { Note, $ED now redirects to U+03D5 as identified by IBM;
          see Wikipedia note. https://bit.ly/39t1732 Better still
          would be U+1D719, but then the table would no longer be
          smallint.}
        cp437ToUnicode: array[0 .. 255] of smallint = (
            $0020, $263A, $263B, $2665, $2666, $2663, $2660, $2022,
            $25D8, $25CB, $25D9, $2642, $2640, $266A, $266B, $263C,
            $25BA, $25C4, $2195, $203C, $00B6, $00A7, $25AC, $21A8,
            $2191, $2193, $2192, $2190, $221F, $2194, $25B2, $25BC,

            $0020, $0021, $0022, $0023, $0024, $0025, $0026, $0027,
            $0028, $0029, $002A, $002B, $002C, $002D, $002E, $002F,
            $0030, $0031, $0032, $0033, $0034, $0035, $0036, $0037,
            $0038, $0039, $003A, $003B, $003C, $003D, $003E, $003F,
            $0040, $0041, $0042, $0043, $0044, $0045, $0046, $0047,
            $0048, $0049, $004A, $004B, $004C, $004D, $004E, $004F,
            $0050, $0051, $0052, $0053, $0054, $0055, $0056, $0057,
            $0058, $0059, $005A, $005B, $005C, $005D, $005E, $005F,
            $0060, $0061, $0062, $0063, $0064, $0065, $0066, $0067,
            $0068, $0069, $006A, $006B, $006C, $006D, $006E, $006F,
            $0070, $0071, $0072, $0073, $0074, $0075, $0076, $0077,
            $0078, $0079, $007A, $007B, $007C, $007D, $007E, $2302,

            $00C7, $00FC, $00E9, $00E2, $00E4, $00E0, $00E5, $00E7,
            $00EA, $00EB, $00E8, $00EF, $00EE, $00EC, $00C4, $00C5,
            $00C9, $00E6, $00C6, $00F4, $00F6, $00F2, $00FB, $00F9,
            $00FF, $00D6, $00DC, $00A2, $00A3, $00A5, $20A7, $0192,
            $00E1, $00ED, $00F3, $00FA, $00F1, $00D1, $00AA, $00BA,
            $00BF, $2310, $00AC, $00BD, $00BC, $00A1, $00AB, $00BB,
            $2591, $2592, $2593, $2502, $2524, $2561, $2562, $2556,
            $2555, $2563, $2551, $2557, $255D, $255C, $255B, $2510,
            $2514, $2534, $252C, $251C, $2500, $253C, $255E, $255F,
            $255A, $2554, $2569, $2566, $2560, $2550, $256C, $2567,
            $2568, $2564, $2565, $2559, $2558, $2552, $2553, $256B,
            $256A, $2518, $250C, $2588, $2584, $258C, $2590, $2580,
            $03B1, $00DF, $0393, $03C0, $03A3, $03C3, $00B5, $03C4,
            $03A6, $0398, $03A9, $03B4, $221E, $03D5, $03B5, $2229,
            $2261, $00B1, $2265, $2264, $2320, $2321, $00F7, $2248,
            $00B0, $2219, $00B7, $221A, $207F, $00B2, $25A0, $0020);
    var
        UnicodeToCP437: array[0 .. 65536] of char;

    function UTF8Len(codepoint: smallint): integer;
	function UTF8ToCodepoint(var utf: array of byte): longint;
	procedure SetupCodepointToCP437;
    function CodepointToCP437(cp: longint): char;
    function CP437ToCodepoint(c: byte): smallint;

implementation

function UTF8Len(codepoint: smallint): integer;
    begin
        if codepoint < $80 then utf8Len := 1
        else if codepoint < $0800 then utf8Len := 2
        else utf8Len := 3
    end;

{ https://rosettacode.org/wiki/UTF-8_encode_and_decode#PureBasic }
function UTF8ToCodepoint(var utf: array of byte): longint;
    begin
        if utf[0] <= %01111111 then UTF8ToCodepoint := Ord(utf[0])
        else if utf[0] <= %11011111 then begin { 2 byte encoding }
            UTF8ToCodepoint :=
                ((Ord(utf[0]) and %00011111) shl 6) or
                ((Ord(utf[1]) and %00111111));
        end else if utf[0] <= %11101111 then begin {3 byte encoding}
            UTF8ToCodepoint :=
                ((Ord(utf[0]) and %00001111) shl 12) or
                ((Ord(utf[1]) and %00111111) shl 6) or
                ((Ord(utf[2]) and %00111111));
        end else if utf[0] <= %11110111 then begin {4 byte encoding}
            UTF8ToCodepoint :=
                ((Ord(utf[0]) and %00000111) shl 18) or
                ((Ord(utf[1]) and %00111111) shl 12) or
                ((Ord(utf[2]) and %00111111) shl 6) or
                ((Ord(utf[3]) and %00111111));
        end else
            UTF8ToCodePoint := 0 {unknown}
    end;

procedure SetupCodepointToCP437;
    var
        i : integer;
    begin
        { Everything that's not known maps to the character '?' }
        fillChar(UnicodeToCP437, 65536, '?');
        { First reverse the cp437 to unicode codepoint table. }
        for i := 0 to 255 do
            UnicodeToCP437[cp437ToUnicode[i]] := Chr(i);

        { Then add some "near-equivalents" }
        {lowercase ø: make it look like phi}
        UnicodeToCP437[248] := Chr(237);

        {uppercase Ø: make it look like Theta}
        UnicodeToCP437[216] := Chr(233);
    end;

function CodepointToCP437(cp: longint): char;
    begin
        if cp > 65535 then CodepointToCP437 := '?'
        else CodepointToCP437 := UnicodeToCP437[cp];
    end;

function CP437ToCodepoint(c: byte): smallint;
    begin
        CP437ToCodepoint := cp437ToUnicode[Ord(c)];
    end;

end.

File: csrc/unicode.cxx
/*
        Copyright (c) 2020 Kristofer Munsterhjelm

        Permission is hereby granted, free of charge, to any person obtaining a copy
        of this software and associated documentation files (the "Software"), to deal
        in the Software without restriction, including without limitation the rights
        to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
        copies of the Software, and to permit persons to whom the Software is
        furnished to do so, subject to the following conditions:

        The above copyright notice and this permission notice shall be included in all
        copies or substantial portions of the Software.

        THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
        IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
        FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
        AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
        LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
        OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
        SOFTWARE.
*/

/* So as not to have to pull in a whole mathematics library just for min and
  max. */

#include "unicode.h"
#include "ptoc.h"
#include <stdexcept>

wchar_t cp437ToUnicode[] = {
    0x20, 0x263a, 0x263b, 0x2665, 0x2666, 0x2663, 0x2660, 0x2022,
    0x25d8, 0x25cb, 0x25d9, 0x2642, 0x2640, 0x266a, 0x266b, 0x263c,
    0x25ba, 0x25c4, 0x2195, 0x203c, 0xb6, 0xa7, 0x25ac, 0x21a8,
    0x2191, 0x2193, 0x2192, 0x2190, 0x221f, 0x2194, 0x25b2, 0x25bc,

    0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27,
    0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
    0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
    0x38, 0x39, 0x3a, 0x3b, 0x3c, 0x3d, 0x3e, 0x3f,
    0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47,
    0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f,
    0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57,
    0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
    0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67,
    0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f,
    0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
    0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x2302,

    0xc7, 0xfc, 0xe9, 0xe2, 0xe4, 0xe0, 0xe5, 0xe7,
    0xea, 0xeb, 0xe8, 0xef, 0xee, 0xec, 0xc4, 0xc5,
    0xc9, 0xe6, 0xc6, 0xf4, 0xf6, 0xf2, 0xfb, 0xf9,
    0xff, 0xd6, 0xdc, 0xa2, 0xa3, 0xa5, 0x20a7, 0x192,
    0xe1, 0xed, 0xf3, 0xfa, 0xf1, 0xd1, 0xaa, 0xba,
    0xbf, 0x2310, 0xac, 0xbd, 0xbc, 0xa1, 0xab, 0xbb,
    0x2591, 0x2592, 0x2593, 0x2502, 0x2524, 0x2561, 0x2562, 0x2556,
    0x2555, 0x2563, 0x2551, 0x2557, 0x255d, 0x255c, 0x255b, 0x2510,
    0x2514, 0x2534, 0x252c, 0x251c, 0x2500, 0x253c, 0x255e, 0x255f,
    0x255a, 0x2554, 0x2569, 0x2566, 0x2560, 0x2550, 0x256c, 0x2567,
    0x2568, 0x2564, 0x2565, 0x2559, 0x2558, 0x2552, 0x2553, 0x256b,
    0x256a, 0x2518, 0x250c, 0x2588, 0x2584, 0x258c, 0x2590, 0x2580,
    0x3b1, 0xdf, 0x393, 0x3c0, 0x3a3, 0x3c3, 0xb5, 0x3c4,
    0x3a6, 0x398, 0x3a9, 0x3b4, 0x221e, 0x3d5, 0x3b5, 0x2229,
    0x2261, 0xb1, 0x2265, 0x2264, 0x2320, 0x2321, 0xf7, 0x2248,
    0xb0, 0x2219, 0xb7, 0x221a, 0x207f, 0xb2, 0x25a0, 0xa0};

std::map<wchar_t, unsigned char> UnicodeToCP437;

void SetupCodepointToCP437() {

    /* First reverse the cp437 to unicode codepoint table. */
    for(int i = 0; i <= 255; ++i) {
        UnicodeToCP437[cp437ToUnicode[i]] = chr(i);
    }

    /* Then add some "near-equivalents" */
    /*lowercase ø: make it look like phi*/
    UnicodeToCP437[248] = chr(237);

    /*uppercase Ø: make it look like Theta*/
    UnicodeToCP437[216] = chr(233);
}

unsigned char CodepointToCP437(wchar_t cp) {
    char CodepointToCP437_result;

    if (UnicodeToCP437.size() == 0) {
        throw std::logic_error("Tried to convert unicode to CP437 without initializing!");
    }

    // Everything that's not known maps to the character '?'
    if (UnicodeToCP437.find(cp) == UnicodeToCP437.end()) {
        return '?';
    }

    return UnicodeToCP437.find(cp)->second;
}

wchar_t CP437ToCodepoint(unsigned char c) {
    return cp437ToUnicode[c];
}


